### PR TITLE
Stock price prediction using Python

### LINKED ISSUE #724: Stock prices prediction using Python
### Brief Summary: 
“Fintech” is one of the most prominent use cases of machine learning.
In this tutorial, I'm going to build a neural network model (LSTM) to predict stock prices with high accuracy.

### Key Takeaways: 
- What is LSTM.
- Choosing data.
- Building the long short term memory model.
- Training the model on our data.
- Plotting results ( I will refer to my "Getting started with data visualization" tutorial for those who don't know it.)

### References: 
N/A

### PR DESCRIPTION
**Pull request by ahmadmardeni1**
#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.

**NOTE: (1st time contributors - submit ONLY one article to get familiar with the process and style-once published, you can then submit any number of articles.)**
      
#### Grammar Checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical Checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution Guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**

This fixes #724

### CODE REVIEW

**Comment by ahmadmardeni1** [comment]
This fixes #724

**Comment by hectorkambow** [comment]
👋 Thank you for submitting this PR - we will review it ASAP - all PRs go into our queue will be looked over before being approved to be published.

**Comment by hectorkambow** [comment]
Plagiarism check done

**Comment by adrianmurage** [comment]
@lalith1403 
Could you kindly take over this review

**Comment by lalith1403** [comment]
I will review this by tomorrow. Sorry for the delay.

**Comment by ahmadmardeni1** [comment]
> I will review this by tomorrow. Sorry for the delay.

No problem, take ur time.

File: articles/stock-price-prediction-using-python/index.md [hunk 1]
```diff
@@ -0,0 +1,170 @@
+# stocks-prices-predection-using-python
+According to [wikipedia](https://en.wikipedia.org/wiki/Artificial_neural_network#:~:text=An%20artificial%20neural%20network%20is,to%20the%20input%20of%20another.) an artificial neural network is an interconnected group of nodes, inspired by a simplification of neurons in a brain. Here, each circular node represents an artificial neuron and an arrow represents a connection from the output of one artificial neuron to the input of another.
```
Comment (lalith1403) [inline articles/stock-price-prediction-using-python/index.md:2]: This definition is not according to Wikipedia. It is a general definition.

File: articles/stock-price-prediction-using-python/index.md [hunk 2]
```diff
@@ -0,0 +1,170 @@
+# stocks-prices-predection-using-python
+According to [wikipedia](https://en.wikipedia.org/wiki/Artificial_neural_network#:~:text=An%20artificial%20neural%20network%20is,to%20the%20input%20of%20another.) an artificial neural network is an interconnected group of nodes, inspired by a simplification of neurons in a brain. Here, each circular node represents an artificial neuron and an arrow represents a connection from the output of one artificial neuron to the input of another.
+
+In this tutorial we are going to build an ai neural network model to predict the stock price. Specifically **Tesla** stock, hope that we can make Elon Mask happy along the way.
```
Comment (lalith1403) [inline articles/stock-price-prediction-using-python/index.md:4]: Musk?

File: articles/stock-price-prediction-using-python/index.md [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+
+In this tutorial we are going to build an ai neural network model to predict the stock price. Specifically **Tesla** stock, hope that we can make Elon Mask happy along the way.
+
+### Prerequisites
+- A good understanding of Python programming language.
+- A good understanding of neural networks, you can start from [here](https://www.section.io/engineering-education/introduction-to-neural-networks/).
+
+#### Development Environment
+- Locally install Python and Tensorflow.
+- Google Colab [Recommended].
```
Comment (lalith1403) [inline articles/stock-price-prediction-using-python/index.md:12]: Share the link to your notebook with all the code and output in it.

File: articles/stock-price-prediction-using-python/index.md [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+from keras.models import Sequential, load_model
+from keras.layers import LSTM, Dense, Dropout
+```
+We are going to use `numpy` for scientific operations, `pandas` to modify our dataset, `matplotlib` to visualize the results, `sklearn` to scale our data, and `keras` works as a wrapper to low-level libraries like TensorFlow or Theano high-level neural networks library.
+
+### Let's build our app!
+First of all, if you take a look at the dataset you need to know that "open" column represents the opening price for the stock at that "date" column, and the "close" column is the closing price on that day. "High" column represents the highest price reached that day, and "Low" column represents the lowest price.
+
+#### Loading data
+Let's upload our dataset(please ignore the following code if you are not using Google Colab):
```
Comment (lalith1403) [inline articles/stock-price-prediction-using-python/index.md:52]: If one is not using Colab, then show the code to load the file.

File: articles/stock-price-prediction-using-python/index.md [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+```python
+from google.colab import files
+dataset = files.upload()
+```
+Then press "choose files" and upload the dataset.
```
Comment (lalith1403) [inline articles/stock-price-prediction-using-python/index.md:57]: You have missed the line... `pd.read_csv('TSLA.csv')`

File: articles/stock-price-prediction-using-python/index.md [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+```
+We will use the `MinMaxScaler` to scale our data between zero and one. In simple words, scaling is converting the numerical data which are represented in a wide range into a smaller one.
+
+```python
+scaler = MinMaxScaler(feature_range=(0,1))
+dataset_train = scaler.fit_transform(dataset_train)
+dataset_test = scaler.transform(dataset_test)
+```
+
+Next, we will create the function which will help us to create the datasets:
```
Comment (lalith1403) [inline articles/stock-price-prediction-using-python/index.md:86]: Explain this function. What is `df[i-50:i,0]`

File: articles/stock-price-prediction-using-python/index.md [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+```python
+x_train, y_train = create_dataset(dataset_train)
+x_test, y_test = create_dataset(dataset_test)
+ ```
+Next, we need to reshape our data to make it a 3D array in order to use it in LSTM Layer.
+```python
+x_train = np.reshape(x_train, (x_train.shape[0], x_train.shape[1], 1))
+x_test = np.reshape(x_test, (x_test.shape[0], x_test.shape[1], 1))
+```
+#### Model building
```
Comment (lalith1403) [inline articles/stock-price-prediction-using-python/index.md:108]: Print model.summary() and include that screenshot. Explain your reasoning behind the architecture.

**Review by lalith1403 — COMMENTED** [review]
Technically, there are a few errors. Grammatically, this article needs more work. The grammar is quite off, and therefore I suggest you find all places that are erroneous and update the same. Use Grammarly, Hemmingway App, or any other service of your choice. Refer to recently published articles to get a feel for the grammatical flow one should follow while writing. 

Give attention to all the suggestions given and make changes accordingly.

File: articles/stock-price-prediction-using-python/index.md (hunk 3 above)
Comment (lalith1403) [reply articles/stock-price-prediction-using-python/index.md:12]: https://colab.research.google.com/github/ahmadmardeni1/Stock-price-predection-using-Python/blob/main/stock_price_predection.ipynb#scrollTo=QI4PhD_bi9dN

**Comment by ahmadmardeni1** [comment]
Good morning @lalith1403 
I did the changes you requested. Please let me know if there is anything else to improve.

File: articles/stock-price-prediction-using-python/index.md [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+author: ahmad-mardeni
+date: 2020-12-30T00:00:00-12:00
+topics: [languages]
+excerpt_separator: <!--more-->
+images:
+
+  - url: /engineering-education/stock-price-prediction-using-python/hero.png
+    alt: stock price prediction python
+---
+An artificial neural network is an interconnected group of nodes, inspired by the brain's simplification. Here, each circular node represents an artificial neuron, and an arrow represents a connection from the output of one artificial neuron to the input of another.
```
Comment (ninjaginja) [inline articles/stock-price-prediction-using-python/index.md:17]: @ahmadmardeni1 are you referencing an image here? If so, the image is missing. If not, I'm not sure what you're referring to when you say, "Here, each circular node represents... and an arrow represents..."

Comment (ninjaginja) [reply articles/stock-price-prediction-using-python/index.md:17]: Also, this is a strange way to begin the article. I'd suggest moving this paragraph down to the neural network section (i.e. under the 'What is RNN?' heading).

Comment (ahmadmardeni1) [reply articles/stock-price-prediction-using-python/index.md:17]: @ninjaginja Hello, it is the Wikipedia definition for the artificial neural network. I will move it.

Comment (ninjaginja) [reply articles/stock-price-prediction-using-python/index.md:17]: In that case, you also need to reference it with a link. Any text that you copy from another source must be properly cited in order to adhere to plagiarism guideline. Please make sure to cite any additional sources/images that may be missing.

Comment (ahmadmardeni1) [reply articles/stock-price-prediction-using-python/index.md:17]: @ninjaginja Hello, it is the Wikipedia definition for the artificial neural network. I don't think we can move it down because we are talking about recurrent neural network there. I will delete it.

Comment (lalith1403) [reply articles/stock-price-prediction-using-python/index.md]: Instead of saying, According to Wikipedia, an RNN is defined as follows, mention the definition and insert the link. I pointed that out because Wikipedia didn't define RNN. Saying, according to Wikipedia, would be erroneous.
